paddle/fluid/framework/details/build_strategy.cc

```diff
@@ -47,212 +47,195 @@ class ParallelExecutorPassBuilder : public ir::PassBuilder {
-      VLOG(1) << "Add graph_viz_pass";
-      auto viz_pass = AppendPass("graph_viz_pass");
-      const std::string graph_path = string::Sprintf(
-          "%s%s", strategy_.debug_graphviz_path_.c_str(), "_original_graph");
-      viz_pass->Set<std::string>("graph_viz_path", new std::string(graph_path));
-    }
+    ResolveOptionConfliction();
 
-    // Note(zcd): record_skip_memory_opt_vars_pass should be the first pass.
-    VLOG(1) << "Add record_skip_memory_opt_vars_pass";
+    AppendPrintGraphPass("graph_viz_pass", "_original_graph");
+    // Note(zcd): record_skip_memory_opt_vars_pass should
+    // be the first pass.
     AppendPass("record_skip_memory_opt_vars_pass");
+    AppendPassWithCheck(strategy_.enable_sequential_execution_,
+                        "sequential_execution_pass");
+    AppendPassWithCheck(strategy_.sync_batch_norm_, "sync_batch_norm_pass");
+
+    AppendOpFusePasses();
+    AppendPrintGraphPass("graph_viz_pass", "_fused_graph");
+    // TODO(dev-paddle): memory optimize pass should be placed last.
+    AppendMemoryOptimizePasses();
+    AppendMultiDevPass();
+    AppendMultiGraphOptPasses();
+
+    AppendPassToSetMkldnnAttr("mkldnn_placement_pass");
+    // runtime_context_cache pass should be the last pass to enable the attr of
+    // all original and fused operators. But no operators can be enabled this
+    // attr if putting it after MultiDevPass.
+    AppendPassWithCheck(strategy_.cache_runtime_context_,
+                        "runtime_context_cache_pass");
+    AppendPassWithCheck(strategy_.remove_unnecessary_lock_,
+                        "modify_op_lock_and_record_event_pass");
+    // Note: This pass is used to check whether the multi_device_graph is right.
+    AppendPass("multi_devices_check_pass");
 
-#ifdef PADDLE_WITH_MKLDNN
-    if (FLAGS_use_mkldnn) {
-      VLOG(1) << "Add mkldnn_placement_pass";
-      AppendPass("mkldnn_placement_pass");
-    } else if (!strategy_.mkldnn_enabled_op_types_.empty()) {
-      LOG(WARNING)
-          << "mkldnn_enabled_op_types specify the operator type list to "
-             "use MKLDNN acceleration. It is null in default, means "
-             "that all the operators supported by MKLDNN will be "
-             "accelerated. And it should not be set when "
-             "FLAGS_use_mkldnn=false.";
-    }
-#else
-    PADDLE_ENFORCE(!FLAGS_use_mkldnn,
-                   "Please compile with MKLDNN first to use MKLDNN");
-#endif
+    SetCollectiveContext();
+  }
 
-    if (strategy_.enable_sequential_execution_) {
-      VLOG(1) << "Add sequential_execution_pass";
-      AppendPass("sequential_execution_pass");
+  void ResolveOptionConfliction() {
+    // Specifies the restrictions between different pass.
+    if (strategy_.enable_parallel_graph_) {
+      VLOG_IF(3, strategy_.fuse_all_optimizer_ops_)
+          << "Currently, fuse_all_optimizer_ops doesn't works under "
+             "parallel_graph.";
+      strategy_.fuse_all_optimizer_ops_ = false;
```

Comment:
If we changed the user's config, there should be LOG(WARNING) and the other under this line is the same problem.

Reply:
I agree with you, so we should recognize whether this option is enabled by the user or not. I will fix this in the next PR https://github.com/PaddlePaddle/Paddle/pull/18833.